src/suggestions.ts

```diff
@@ -0,0 +1,120 @@
+import * as fs from 'fs';
+import {PathLike} from 'fs';
+import * as path from 'path';
+import * as github from '@actions/github';
+
+import * as sorald from './sorald';
+import {Range} from './ranges';
+import * as git from './git';
+import {Hunk, Repo} from './git';
+
+/**
+ * A patch suggestion meant to be posted to GitHub usuing the pull request
+ * review comment API.
+ */
+export interface PatchSuggestion {
+  linesToReplace: Range;
+  file: PathLike;
+  suggestion: string;
+  violationSpec: string;
+}
```

Comment:
The purpose of an interface is to achieve abstraction and create a blueprint for classes which shall implement this but you just have to use it as a type. And for that TypeScript has created an eponymous language construct called [`type`](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-aliases).

The documentation says that it [hardly makes a difference](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces) but I would simply use `type` because it is made solely for what you want to achieve.

Reply:
> The purpose of an interface is to achieve abstraction and create a blueprint for classes which shall implement this but you just have to use it as a type.

Can you name one use of an interface where this is _not_ the case?

Note that `type` in TypeScript is a _type alias_, and not the definition of "type" (which is much broader, e.g. a class is a type, an interface is a type, etc). I think there are two strong reasons to use an interface over a type alias here.

1. Interfaces are all but ubiquitous in object-oriented languages, and the concept of an interface generalizes far to almost all aspects of computer programming (think e.g. API), and even beyond. Most engineers, programmers or not, understand the concept of an interface. Type aliases, while present in many languages, are far less recognizable (notably, they do not exist in Java). 
2. The documentation you linked suggests as much: "If you would like a heuristic, use interface until you need to use features from type."

There's no need for any of the specific features from `type` here, so it's imo better to use the more recognizable `interface`.

Personally, I'd use type aliases primarily to ease readability and reduce duplication. For example, say that I use the union type `string | number` as some form of identifier, then I can alias that to `type Identifier = string | number` and use that. That's something that `interface` can't do, and so `type` steps in to fill the gap.

Reply:
>> The purpose of an interface is to achieve abstraction and create a blueprint for classes which shall implement this but you just have to use it as a type.

> Can you name one use of an interface where this is not the case?

I was not supporting the usage of interface for this scenario as you are just concerned with the _structure_ of the POST body. TypeScript has `type` constructs to define the structure. Otherwise, for all other languages like Java, C++, I support usage of interface and struct respectively.

> Interfaces are all but ubiquitous in object-oriented languages, and the concept of an interface generalizes far to almost all aspects of computer programming (think e.g. API), and even beyond. Most engineers, programmers or not, understand the concept of an interface. Type aliases, while present in many languages, are far less recognizable (notably, they do not exist in Java).

Very strong reason indeed. I was thinking exactly the opposite. My mind said if TypeScript has implemented `type` for when we are only concerned with the structure, let's put it to use where it fits. 😅 

So to conclude this, I think type aliases are best used when we have to combine types for an entity using union or intersection operations. For scenarios where we need to provide structure, we can simply use `interface` because of your reasons above and the documentation link.

Reply:
> My mind said if TypeScript has implemented type for when we are only concerned with the structure

I disagree, and I think you have it a bit backwards. TypeScript is structurally typed, so interfaces are by definition only specifications of structure. That is inherent in the type system. Here's another quote from the documentation you linked to further strengthen that notion:

> Interfaces may only be used to declare the shapes of objects [...]

In other words, interfaces can _only_ declare structure, and nothing more.

`type` is a _type alias_, which _can_ define structure just like an interface, but that's not it's only purpose. For example, it can define union types (e.g. `type Identifier = string | number`), it can be used to just _rename_ a type (e.g. `type StRiNg = string;`, hence the _alias_ in the name), and some other interesting things.

Reply:
> so interfaces are by definition only specifications of structure. That is inherent in the type system.

Got it. Thanks! I wrongly interpreted the documentation when it said:
> Just like when we used a type alias above, the example works just as if we had used an anonymous object type. TypeScript is only concerned with the structure of the value we passed to printCoord - it only cares that it has the expected properties. Being concerned only with the structure and capabilities of types is why we call TypeScript a structurally typed type system.

> which can define structure just like an interface

Two ways to do the same thing confuses me lot. But I think it's clear, for now.

Reply:
> Two ways to do the same thing confuses me lot. But I think it's clear, for now.

I agree, `type` and `interface` are _deceptively_ similar in TypeScript. If you search for it, there are tons of articles of "when to use which", which kind of indicates that they are too similar. IMO, it would have been better to keep type aliases to only be just that, aliases, whereas in TypeScript they do more than just aliasing (e.g. they can for all intents and purposes define interfaces).

Reply:
> If you search for it, there are tons of articles of "when to use which", which kind of indicates that they are too similar.

Yeah, the differences are so subtle that they can't even be perceived.